clar2wasm/tests/wasm-generation/principal.rs

```diff
@@ -0,0 +1,266 @@
+        let principal_data: [u8; 20] = principal
+            .try_into()
+            .expect("slice must be exactly 20 bytes");
+
+        match contract_name {
+            Some(contract_name) => {
+                Value::Principal(PrincipalData::Contract(QualifiedContractIdentifier::new(
+                    StandardPrincipalData(version, principal_data),
+                    contract_name.into(),
+                )))
+            }
+            None => Value::Principal(PrincipalData::Standard(StandardPrincipalData(
+                version,
+                principal_data,
+            ))),
+        }
+    }
+
+    fn create_error_construct(error_code: u8, principal_data: Option<Value>) -> Value {
+        Value::error(
+            TupleData::from_data(vec![
+                ("error_code".into(), Value::UInt(error_code.into())),
+                (
+                    "value".into(),
+                    Value::Optional(OptionalData {
+                        data: principal_data.map(Box::new),
+                    }),
+                ),
+            ])
+            .unwrap()
+            .into(),
+        )
+        .unwrap()
+    }
+
+    fn create_error_destruct(
+        hash_bytes: Value,
+        version_byte: u8,
+        data: Option<Box<Value>>,
+    ) -> Value {
+        Value::error(
+            TupleData::from_data(vec![
+                ("hash-bytes".into(), hash_bytes),
+                ("name".into(), Value::Optional(OptionalData { data })),
+                (
+                    "version".into(),
+                    Value::Sequence(SequenceData::Buffer(BuffData {
+                        data: vec![version_byte],
+                    })),
+                ),
+            ])
+            .unwrap()
+            .into(),
+        )
+        .unwrap()
+    }
+
+    fn key_to_stacks_addr(address_version: u8, key: &StacksPrivateKey) -> StacksAddress {
+        StacksAddress::from_public_keys(
+            address_version,
+            &AddressHashMode::SerializeP2PKH,
+            1,
+            &vec![StacksPublicKey::from_private(key)],
+        )
+        .unwrap()
+    }
+
+    proptest! {
+        #![proptest_config(runtime_config())]
+
+        #[test]
+        fn crosscheck_principal_construct(
+            version_byte in strategies_for_version_byte(),
+            hash_bytes in buffer(20),
+            contract in option::of(string_regex("([a-zA-Z](([a-zA-Z0-9]|[-])){0,30})").unwrap())
+        ) {
+            let expected_principal = create_principal(
+                version_byte as u8,
+                &hash_bytes.clone().expect_buff(20).unwrap(),
+                contract.as_deref(),
+            );
+
+            let expected = match version_byte {
+                // version_bytes for Mainnet: single_sig (0x14) and multi_sig (0x16).
+                // version_bytes for Testnet: single_sig (0x1A) and multi_sig (0x15).
+                0x1A | 0x14 | 0x15 | 0x16 => Value::okay(expected_principal),
```

Comment:
That misses tests with version bytes for mainnet on Testnet and for version bytes of Testnet on Mainnet.